Feature/threading

krowvin opened the pull request:
Added optional flag `-p` or `--parallel` to enable multi-threaded CDA fetch requests. This can drop run times significantly since CDA requests are I/O bound and not CPU bound (i.e. waiting on the database/network). 

## What
- Use queue/threading and split the CDA logic out for parallel requests
- revert to original run method if the `-p` flag is not provided
- Add a pytest for fetching using the `-p` flag and compares against an expected output (set datetime) / Add this to the github workflow
- Split the test directories out by district, thinking more districts/we can create per district tests for active forms
- Move various variables like version, thread count, repgen5 github issues url to the `repgen/__init__.py`
- Experiment with how to go about looping projects and dynamically setting the variables for form generation

## Why
- SWT has a monthly report that takes upwards of 5 minutes to complete. This won't matter as much when things are running headless/on cron/airflow/etc. But during testing having the ability to run tests in parallel for a complete report output in a fraction of the time will be great!


# Bumped version to `5.1.0`



# Issue(s) #
- #30

DanielTOsborne commented on repgen/data/value.py:
```diff
@@ -400,182 +401,8 @@ def parse_slice(value):
 			except Exception as err:
 				print( repr(err) + " : " + str(err), file=sys.stderr )
 		elif self.dbtype.upper() in ["JSON", "RADAR"]:
```
Since you change this code already, go ahead and add "CDA" as a dbtype option. I meant to add it myself.

krowvin replied:
There is a separate PR here for this

- #32

DanielTOsborne commented on repgen/__init__.py:
```diff
@@ -0,0 +1,11 @@
+# =========================================== #
+#              THREADING VARS				  #
+# =========================================== #
+THREAD_COUNT = 50
```
I'm concerned this is too high. Spawning threads has overhead. In addition, python has a global interpreter lock which prevents true parallelization.
Plus, that number of connections could cause the OS near the TCP limit to fail to open more, or the remote server could throttle/block more than a certain number.

I'd reduce this to maybe 10 or 20 at most, see how it performs.

krowvin replied:
Using the `bulk_test.frm` which does 48 projects (timeseries) for 24 hours of hourly data. 

| Threads | Time (s) |
| -------- | ------- |
| 1 | 22.36    |
| 5| 4.79    |
| 10 | 3.55    |
| 15 | 1.85    |
| 20 | 1.78    |
| 50 | 1.36    |

I think 15 threads is reasonably in the middle of 10 or 20 and the results seem to diminish for my very limited testing. 

I do not expect this to scale linearly and results will vary with more/less timeseries and network conditions. 

I made sure to re-run each test multiple times to rule out CDA DB caching.

DanielTOsborne commented on repgen/data/value.py:
```diff
@@ -57,7 +64,6 @@ class Value:
 	}
 
-	# This isn't thread safe, not an issue yet though since repgen isn't multithreaded.
 	_conn = None
```
Is this still used? The comment was removed, but the variable itself wasn't. If this is still used, is it threadsafe now?

krowvin replied:
The variable is a `_` meaning it should be private if such a scope existed for python

The way I did the threads each of the Value classes spawns it's own call for data. This means the scope is limited for threading to each Value class. 

This was just an oversight after all those changes and I missed this comment.

DanielTOsborne commented on repgen/data/value.py:
```diff
@@ -1183,7 +1022,7 @@ def min( *args, **kwarg ):
 						if isinstance( treat, number_types):
 							tmp.value = treat
-						elif treat=="MISS":
+						elif py=="MISS":
```
Why was this changed, what is 'py'?

krowvin replied:
I think, this was part of a global rename! This needs to be fixed. Although I can't fathom why I would have done a replace for the word treat?

DanielTOsborne commented on repgen/report/report.py:
```diff
@@ -123,14 +125,19 @@ def run( self, basedate, local_vars: dict = None ):
 
 		for key in my_locals:
 			if isinstance(my_locals[key], Value ):
+				# TODO: this could cause a bug if someone is placing a % in the report with the same text after it i.e. css
```
I'd like more information on this bug.

krowvin replied:
This probably should be it's own issue to be honest, but I left the todo in there as a 1-off if someone ever went looking for the bug. 

But lets say you have a Value such as

*(arguably could be named better)*
```python
myVar = Value()
```

And then in an HTML file (form) you are creating you reference your value in the text with 
```html
<td>%myVar</td>
```

But perhaps further down you have something like this
```html
<script>
  const myVar = 5;
  // Is divisible by myVar with a remainder?
  console.log(otherTotal%myVar)
  // Could be spaced better..
</script>
```
*Would* this hardcode the value into the form and force it to throw an error during the javascript runtime?

# The main reason this was a TODO 
Was because I thought about it, but I don't think it's so urgent it needs to be an active issue. Because **what** are the chances?

DanielTOsborne commented on repgen/workers/http.py:
```diff
@@ -0,0 +1,226 @@
+						# Test if the connection is valid
+						v_self._conn.getresponse().read()
+					
+					# locals().update({'max_call_size': locals().get('max_call_size', 0) + 1})
```
What's this, it seems incomplete?

krowvin replied:
This was an attempt at adding a forced upper ceiling. In my opinion we might consider setting a MAX for a report. 

Prior to this the max was how long they want to wait. I assume AirFlow also has the ability to kill a job running for too long. 

But what if the job is threaded now and can fire 50 (or less) threads and ends up in an infinite loop? 
With a max call size, threaded or not, we could throw and error and provide a flag to override. Or just set it to something so large people should never hit it in one report. (5000, 10000?)
( I set it to [this](https://github.com/USACE-WaterManagement/repgen5/pull/35/files#diff-eae2fa579373c57cbe59cd0082bf77089a8881d8cab1caed7c26f21aaf669f8aR7) for initial testing and ended up commenting this out)

DanielTOsborne commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
-python-dateutil==2.8.2
-pytz==2024.1
+coverage==5.4
+pytz==2022.1
```
Why was pytz downgraded?

krowvin replied:
I was running the freeze in the wrong environment and this is me reverting those extra changes. 

I also noticed there were at least two requirements files in the project. Had to make sure not to overwrite those or get them confused. 

Might we work checking to see what the version was prior because the line is not green/red meaning I think that's what it was before? I could be wrong there though

I.e. hasn't been updated since 2022.1 ?

DanielTOsborne commented on tests/spk/testfiles/i-base.battery:
```diff
@@ -105,4 +105,4 @@ UNDEF="????"
 #
 #
-#ENDDEF
+#ENDDEF
```
Your editor removed the EOL marker from a bunch of files.
The EOL marker on the last line of a file is used by some editors to determine the EOL style (CR/CRLF/LF).

krowvin replied:
Ah good catch, is there a setting in VSCode to prevent this? 

Must have happened when I split the files out to make my own test files

DanielTOsborne commented on tests/swt/testfiles/test.frm:
```diff
@@ -0,0 +1,121 @@
+    # Dynamically add the project to the locals for printing to the form
+    locals().update(
+        {
+            project + "-elev": elev
```
While this isn't an issue with code in general, I don't recommend using a dash in the names.
The form labels (%name) is based on actual variable names, and the dash isn't a valid character in a variable name. Use underscore (_) instead.

e.g. `%CANT-elev` is the diff variable subtracted from the CANT variable `CANT - elev`

I'm surprised this worked. 😄

krowvin replied:
Python!!
I figured it was the "key" and didn't even think that it was the variable name within the locals

DanielTOsborne commented on tests/test_converter.py:
```diff
@@ -5,10 +5,10 @@
-EXPECT_DIR = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "expect"))
-ACTUAL_DIR = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "actual"))
-LOG_DIR = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "output"))
+INPUT_DIR = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "spk", "testfiles"))
```
I hadn't intended to maintain separate tests per district. I suppose that's fine though.

krowvin replied:
I made an issue about this

In my opinion we should keep a store of various reports districts have so when we are making major changes we can confirm reports will still function against what are otherwise production reports

- #28

DanielTOsborne commented on __main__.py:
```diff
@@ -37,38 +42,40 @@ def parseArgs():
-        # rejoin the rest:
-        value = '='.join(items[1:])
-    return (key, value)
+	"""
```
Were there any code changes other than changing the indent?
I thought I fixed all the mixed indent in the files. I'm surprised python didn't throw an error when running repgen previously.

krowvin replied:
As far as I know you can actually mix tab styles between scope

`\t` in a def block but then `••••` everywhere else

DanielTOsborne reviewed (commented):
The actual threading code seems mostly fine, aside from one section I added a prior comment to, there's just a lot of unrelated changes and questions.

krowvin commented:
@jbkolze downloaded the branch and did some testing. 

Right out of the gate he realized the threads were not actually doing anything...
From 
https://github.com/USACE-WaterManagement/repgen5/pull/35/commits/d8f04d7ec7911a7a3a627b03b4e38db8d8e01d83
Onward I worked on fixing that issue and was able to get the runtimes down from 20seconds to 1-2seconds for 15 threads. 

However, there is one looming issue. I spent some time trying to solve it. 

# This branch is not production ready without this bug fix
__tldr; race conditions are hard!__

Currently you are able to re-use Value calls, and this seemingly works.

For example if you do:
```python
_value = Value(
        dbtype="radar",
        DBLOC="KEYS",
        DBPAR="Elev",
        DBPTYP="Inst",
        DBINT="1Hour",
        DBDUR="0",
        DBVER="Ccp-Rev",
        DBUNITS="ft",
    )
```
This will write the values to the form output

Then if you chain another and change just the project for example
```python
_value = Value(
        DBLOC="FGIB"
    )
```
Then this will *also* write values the correct values to the form. 

So it seems the [`deepcopy`](https://github.com/USACE-WaterManagement/repgen5/blob/c22a0af62df7cd3d756ce2788153fe2423e5bebe/repgen/data/value.py#L142) code works across the values. 

The problem arises when you want to take the results of some value and feed *that* into another Value constructor. For example:

```python
_value = Value(
        DBLOC="FGIB"
    )
    
_dates = Value(_value.datatimes(),
        PICTURE="%d%b%Y %K%M",
    ) 
```
This will end up failing because `_value` will be in another thread fetching the data and is *non-blocking* in this new threading branch. So it will continue on with an empty `Values` property array that then results in no results for `.datatimes()`. 

Some thoughts I had on how to deal with this:
1. We switch to async io and determine per value if it is a TIMESERIES `type` or a COPY `type` for the Value - If not a TS type we block the request and wait for others (requires end users to be aware of how to structure their forms)
2. We split the string into a third category internally. 
  - Right now we split into the FORM and the DEF. We could split the DEF into a DEP and INDEP (dependent and independent)
   - Then have all the independent values fire first (your timeseries/data calls) 
   - Followed by all the DEP calls. 
   ⚠️ Problem with this method is figuring out how to pass the values after the fact since (afaik) python does not have the concept of [promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). Python's threading [Event](https://docs.python.org/3/library/threading.html#threading.Event) came to mind though.
   
 3. Instead of having the actual data in `datatimes()` we return the properties of the class and have it *refetch* values (inefficient)
 
 
 I'm moving on to the other pressing issues - but will revisit this time willing!

MikeNeilson commented:
Okay, so there are 2 ways (well, maybe more but I can only think of 2) to deal with this chicken and egg problem.
1.
Value itself, while it should be thread safe, is independent and doesn't do any threading.... or maybe it is, like you said this type of design is hard
Instead you make a function that takes a list of Value objects and returns them, this function itself waits, but otherwise retrieves/let's retrieve the Values in parallel.

2. Instead of letting the user make a "Value" directly you use a factory method to provide a "Future/Promise" (whatever it's called in Python) and when you get to a portion that has to wait you can get the actual Value.


After typing that, they aren't really mutually exclusive. Using more of a builder pattern (factory method, actual class, etc) we could just initialize the object with the query parameters and another call is required to actually get them.

However, consider that will add some complexity to people getting onboarded. Honestly I think I like the idea of method that takes multiple definitions and when it returns the data is ready. Seems like explaining "set all your data, then do math" is easier to explain than the intricacies of threading issues.

krowvin commented:
> sing more of a builder pattern (factory method, actual class, etc) we could just initialize the object with the query parameters and another call is required to actually get them. 

Seems #21  is right in line with what you suggest here.
